### PR TITLE
Add Austrian district data.

### PR DESCRIPTION
**Pull request by geening**
How can I test the changes to eurostat/index.csv?

### CODE REVIEW

File: src/data/metadata.csv
Comment (owahltinez) [inline src/data/metadata.csv:696]: It seems that `(Land)` is typically written as `-Land` ([example](https://en.wikipedia.org/wiki/Sankt_P%C3%B6lten-Land_District)) and that `(Stadt)` is omitted ([example](https://en.wikipedia.org/wiki/Linz)).

I would suggest these changes:
```suggestion
AT_101,AT,Austria,1,Burgenland,101,Eisenstadt,,,,
AT_102,AT,Austria,1,Burgenland,102,Rust,,,,
AT_103,AT,Austria,1,Burgenland,103,Eisenstadt-Umgebung,,,,
AT_104,AT,Austria,1,Burgenland,104,Güssing,,,,
AT_105,AT,Austria,1,Burgenland,105,Jennersdorf,,,,
AT_106,AT,Austria,1,Burgenland,106,Mattersburg,,,,
AT_107,AT,Austria,1,Burgenland,107,Neusiedl am See,,,,
AT_108,AT,Austria,1,Burgenland,108,Oberpullendorf,,,,
AT_109,AT,Austria,1,Burgenland,109,Oberwart,,,,
AT_201,AT,Austria,2,Carinthia,201,Klagenfurt Stadt,,,,
AT_202,AT,Austria,2,Carinthia,202,Villach Stadt,,,,
AT_203,AT,Austria,2,Carinthia,203,Hermagor,,,,
AT_204,AT,Austria,2,Carinthia,204,Klagenfurt Land,,,,
AT_205,AT,Austria,2,Carinthia,205,Sankt Veit an der Glan,,,,
AT_206,AT,Austria,2,Carinthia,206,Spittal an der Drau,,,,
AT_207,AT,Austria,2,Carinthia,207,Villach Land,,,,
AT_208,AT,Austria,2,Carinthia,208,Völkermarkt,,,,
AT_209,AT,Austria,2,Carinthia,209,Wolfsberg,,,,
AT_210,AT,Austria,2,Carinthia,210,Feldkirchen,,,,
AT_301,AT,Austria,3,Lower Austria,301,Krems an der Donau,,,,
AT_302,AT,Austria,3,Lower Austria,302,Sankt Pölten,,,,
AT_303,AT,Austria,3,Lower Austria,303,Waidhofen an der Ybbs,,,,
AT_304,AT,Austria,3,Lower Austria,304,Wiener Neustadt,,,,
AT_305,AT,Austria,3,Lower Austria,305,Amstetten,,,,
AT_306,AT,Austria,3,Lower Austria,306,Baden,,,,
AT_307,AT,Austria,3,Lower Austria,307,Bruck an der Leitha,,,,
AT_308,AT,Austria,3,Lower Austria,308,Gänserndorf,,,,
AT_309,AT,Austria,3,Lower Austria,309,Gmünd,,,,
AT_310,AT,Austria,3,Lower Austria,310,Hollabrunn,,,,
AT_311,AT,Austria,3,Lower Austria,311,Horn,,,,
AT_312,AT,Austria,3,Lower Austria,312,Korneuburg,,,,
AT_313,AT,Austria,3,Lower Austria,313,Krems-Land,,,,
AT_314,AT,Austria,3,Lower Austria,314,Lilienfeld,,,,
AT_315,AT,Austria,3,Lower Austria,315,Melk,,,,
AT_316,AT,Austria,3,Lower Austria,316,Mistelbach,,,,
AT_317,AT,Austria,3,Lower Austria,317,Mödling,,,,
AT_318,AT,Austria,3,Lower Austria,318,Neunkirchen,,,,
AT_319,AT,Austria,3,Lower Austria,319,Sankt Pölten-Land,,,,
AT_320,AT,Austria,3,Lower Austria,320,Scheibbs,,,,
AT_321,AT,Austria,3,Lower Austria,321,Tulln,,,,
AT_322,AT,Austria,3,Lower Austria,322,Waidhofen an der Thaya,,,,
AT_323,AT,Austria,3,Lower Austria,323,Wiener Neustadt-Land,,,,
AT_325,AT,Austria,3,Lower Austria,325,Zwettl,,,,
AT_401,AT,Austria,4,Upper Austria,401,Linz,,,,
AT_402,AT,Austria,4,Upper Austria,402,Steyr,,,,
AT_403,AT,Austria,4,Upper Austria,403,Wels,,,,
AT_404,AT,Austria,4,Upper Austria,404,Braunau am Inn,,,,
AT_405,AT,Austria,4,Upper Austria,405,Eferding,,,,
AT_406,AT,Austria,4,Upper Austria,406,Freistadt,,,,
AT_407,AT,Austria,4,Upper Austria,407,Gmunden,,,,
AT_408,AT,Austria,4,Upper Austria,408,Grieskirchen,,,,
AT_409,AT,Austria,4,Upper Austria,409,Kirchdorf an der Krems,,,,
AT_410,AT,Austria,4,Upper Austria,410,Linz-Land,,,,
AT_411,AT,Austria,4,Upper Austria,411,Perg,,,,
AT_412,AT,Austria,4,Upper Austria,412,Ried im Innkreis,,,,
AT_413,AT,Austria,4,Upper Austria,413,Rohrbach,,,,
AT_414,AT,Austria,4,Upper Austria,414,Schärding,,,,
AT_415,AT,Austria,4,Upper Austria,415,Steyr-Land,,,,
AT_416,AT,Austria,4,Upper Austria,416,Urfahr-Umgebung,,,,
AT_417,AT,Austria,4,Upper Austria,417,Vöcklabruck,,,,
AT_418,AT,Austria,4,Upper Austria,418,Wels-Land,,,,
AT_501,AT,Austria,5,Salzburg,501,Salzburg,,,,
AT_502,AT,Austria,5,Salzburg,502,Hallein,,,,
AT_503,AT,Austria,5,Salzburg,503,Salzburg-Umgebung,,,,
AT_504,AT,Austria,5,Salzburg,504,Sankt Johann im Pongau,,,,
AT_505,AT,Austria,5,Salzburg,505,Tamsweg,,,,
AT_506,AT,Austria,5,Salzburg,506,Zell am See,,,,
AT_601,AT,Austria,6,Styria,601,Graz,,,,
AT_603,AT,Austria,6,Styria,603,Deutschlandsberg,,,,
AT_606,AT,Austria,6,Styria,606,Graz-Umgebung,,,,
AT_610,AT,Austria,6,Styria,610,Leibnitz,,,,
AT_611,AT,Austria,6,Styria,611,Leoben,,,,
AT_612,AT,Austria,6,Styria,612,Liezen,,,,
AT_614,AT,Austria,6,Styria,614,Murau,,,,
AT_616,AT,Austria,6,Styria,616,Voitsberg,,,,
AT_617,AT,Austria,6,Styria,617,Weiz,,,,
AT_620,AT,Austria,6,Styria,620,Murtal,,,,
AT_621,AT,Austria,6,Styria,621,Bruck-Mürzzuschlag,,,,
AT_622,AT,Austria,6,Styria,622,Hartberg-Fürstenfeld,,,,
AT_623,AT,Austria,6,Styria,623,Südoststeiermark,,,,
AT_701,AT,Austria,7,Tyrol,701,Innsbruck-Stadt,,,,
AT_702,AT,Austria,7,Tyrol,702,Imst,,,,
AT_703,AT,Austria,7,Tyrol,703,Innsbruck-Land,,,,
AT_704,AT,Austria,7,Tyrol,704,Kitzbühel,,,,
AT_705,AT,Austria,7,Tyrol,705,Kufstein,,,,
AT_706,AT,Austria,7,Tyrol,706,Landeck,,,,
AT_707,AT,Austria,7,Tyrol,707,Lienz,,,,
AT_708,AT,Austria,7,Tyrol,708,Reutte,,,,
AT_709,AT,Austria,7,Tyrol,709,Schwaz,,,,
AT_801,AT,Austria,8,Vorarlberg,801,Bludenz,,,,
AT_802,AT,Austria,8,Vorarlberg,802,Bregenz,,,,
AT_803,AT,Austria,8,Vorarlberg,803,Dornbirn,,,,
AT_804,AT,Austria,8,Vorarlberg,804,Feldkirch,,,,
AT_900,AT,Austria,9,Vienna,900,Wien,,,,
```

File: src/pipelines/epidemiology/config.yaml
Comment (owahltinez) [inline src/pipelines/epidemiology/config.yaml:120]: Tests probably pass because there's at least one district picked at random from the locations starting with `AT_*`, but to be more precise this should instead be:

```suggestion
      location_key_match: 'AT_[^_]+[^_]+'
```

That way, we can run `update.py --only epidemiology --location-key AT_*_*` and only this data source will run.

**Review by owahltinez — CHANGES_REQUESTED** [review]
This is great! Only a couple of minor comments. You can test the changes to Eurostat by running a data source which uses that data, for example: `update.py --only demographics --location-key AT_*` and then check if the output for the newly added keys has any data.

File: src/data/metadata.csv
Comment (geening) [reply src/data/metadata.csv:696]: Where is subregion2_name actually used?  I'm pulling these names from the Austria data source (though the pipeline matches on a key constructed from the code, and not subregion2_name, so subregion2_name is irrelevant for that purpose).  Will using a more common name help in other ways, such as users searching for it?

Comment (owahltinez) [reply src/data/metadata.csv:696]: Names are used to identify the region in the index as you described (users searching for it). Generally, we prefer the American English names of places but it's not a strict requirement. In all cases, we want to add the keys and the corresponding Wikidata ID to the `knowledge_graph.csv` file so they can be disambiguated without depending on the name.

Comment (geening) [reply src/data/metadata.csv]: > It seems that `(Land)` is typically written as `-Land` ([example](https://en.wikipedia.org/wiki/Sankt_P%C3%B6lten-Land_District)) and that `(Stadt)` is omitted ([example](https://en.wikipedia.org/wiki/Linz)).
> 
> I would suggest these changes:

Okay, I made the changes separately (whenever I said to accept suggestion it said there were recent changes ad to refresh).  FYI the changes didn't affect names that had a space before Land or a space before Stadt -- just confirming that those should stay as they are.

Also, these are not American English names I think, for example this uses Wien instead of Vienna.  Do you happen to know if these districts all have American English names?  We use American English names for the states.

File: src/pipelines/epidemiology/config.yaml
Comment (geening) [reply src/pipelines/epidemiology/config.yaml]: After making this change, I can't get the data source to run with "update.py --only epidemiology --location-key AT_*_*" or with "update.py --only epidemiology --location-key AT_*" -- in both cases, only the level 2 Austria source runs and not the level 3.

File: src/data/metadata.csv
Comment (owahltinez) [reply src/data/metadata.csv]: I normally get the American English names from Wikidata. The nice thing about using Wikidata as the source of truth is that you might also find locations which currently have no data, so they don't appear in the data source, but that can always change in the future.

To get a list of all municipalities in Austria from Wikidata, you'll need to:
1. Find the corresponding Wikidata property you can filter by, I normally go to the Wikipedia page of a location ([example](https://en.wikipedia.org/wiki/Sankt_P%C3%B6lten-Land_District)), then go to the Wikidata item (via the left panel, which takes me [here](https://www.wikidata.org/wiki/Q694393)) and search there for something that fits. In this case, [this](https://www.wikidata.org/wiki/Q871419) is probably what I would use. Unfortunately, I don't see the ID which you are using the code anywhere, so we might need to do manual matching.
2. Write a SPARQL script to query the Wikidata knowledge graph. There are several examples [here](https://github.com/GoogleCloudPlatform/covid-19-open-data/blob/main/src/scripts/queries.sparql) which I normally copy/paste and then change the property IDs.

Querying Wikidata and dealing with SPARQL can be pretty hard. If you get stuck, I can give it a try tomorrow and I'll attach a CSV with all the keys, names and Wikidata IDs.

Sometimes Wikidata does not have the right data (for example missing subregion codes) in which case there's nothing we can do other than manually matching each record or, if we're lucky, use string matching based on the name.

File: src/pipelines/epidemiology/config.yaml
Comment (geening) [inline src/pipelines/epidemiology/config.yaml:120]: Hmm, I thought I commented on this already but don't see it.  When I make this change, 

update.py --only epidemiology --location-key AT_*_*
and
update.py --only epidemiology --location-key AT_*

both only run the level 2 data source.

Comment (owahltinez) [reply src/pipelines/epidemiology/config.yaml:120]: That sounds possible, I think the logic which filters data sources from the location key parameter is still pretty buggy.

Comment (geening) [reply src/pipelines/epidemiology/config.yaml:120]: location_key_match seems to serve two purposes.

It is used in the data source test to match the metadata key in https://github.com/GoogleCloudPlatform/covid-19-open-data/blob/18dbd883b1d18c3d53a8c524aee2154934bea563/src/test/test_source_run.py

And it is used in update.py to filter which pipelines to run, by matching the location-key parameter.

location_key_match: 'AT_[^_]+[^_]+' seems to be fine for matching the metadata key.

However, because the metadata keys for level 2 are of the form 'AT' or 'AT_' followed by a number, and the keys for level 3 are of the form 'AT_' followed by three numbers, any location_key_match that matches every level 2 string (note that the existing 'AT_' does not capture the country-level key 'AT'; is that a problem in itself?) would also match every level 3 string.

Does location_key_match need to capture every metadata key that the source matches?  If so, it's not possible to set up level 2 and level 3 location_key_match strings that allow running only the level 3 source but not the level 2 source.

Perhaps it's okay though to always run the level 2 source along with the level 3 source -- if so, the existing location_key_match is fine; to run the level 3 source, we just need to specify update.py --only epidemiology --location-key AT_** or something else of the sort (since 'AT_*_*' has two underscores, it doesn't match 'AT_[^_]+[^_]+').

File: src/data/metadata.csv
Comment (geening) [reply src/data/metadata.csv]: The best idea I have is combining
https://w.wiki/pJH
and
https://w.wiki/pJK

joining the tables by wikidata id, and then manually matching with the slightly differently formatted names in our datafile.

Any thoughts on an easier way?

Comment (geening) [reply src/data/metadata.csv]: Looks like the German-language query returns names which are usually just the district name preceded by "Bezirk", so I think I'll just preface the names currently in metadata.csv by "Bezirk" and try to join the tables -- that should take care of most of the cases.

Would be great if there were a way to automatically do a wikidata.org search for a list of terms and get the wikidata id of the first result.

Comment (owahltinez) [reply src/data/metadata.csv]: > Any thoughts on an easier way?

Here is what I came up with, inspired by [this](https://stackoverflow.com/a/45417484): https://w.wiki/pLG

> Would be great if there were a way to automatically do a wikidata.org search for a list of terms and get the wikidata id of the first result.

You can query one label at a time... However querying by name is very finicky, it's always much better to search based on a property. That way you also get locations which might not be captured in the data source, which can actually expose bugs in the way we process data and also guarantees that we are respecting a reasonable hierarchical order. Data sources sometimes invent new locations for the sole purpose of COVID-19 data reporting, for example.

File: src/pipelines/epidemiology/config.yaml
Comment (owahltinez) [reply src/pipelines/epidemiology/config.yaml:120]: > location_key_match seems to serve two purposes.

Yes, it's kind of a hack. I was trying to implement an easy way to test newly added data sources without having to comment out all other data sources from the `config.yaml` file.

> Does location_key_match need to capture every metadata key that the source matches? If so, it's not possible to set up level 2 and level 3 location_key_match strings that allow running only the level 3 source but not the level 2 source.

Right, `AT_*` should indeed match levels 2 and 3. It's not even proper RegEx, it should be `AT_.+`. If you can think of a better way to make this work, I'm open to it. Originally I was thinking something like `update.py [...] --country XX --level [0,1,2,3]` and then we can build the proper RegEx automatically without having to type it all out on the command line. Alternatively we can process the `location_key` parameter as a non-regex string and replace `*` with `[^_]+`.

> since 'AT__' has two underscores,

Which location has two underscores? That sounds like a bug.

Comment (geening) [reply src/pipelines/epidemiology/config.yaml:120]: No location has two underscores.  I was using raw text and it removed the star; I meant to say `AT_*_*`,
referring to the location-key parameter you had directed me to use for testing level 3 by itself.

Building the proper RegEx string for a country and level will depend on the format of that country's keys, so I don't see how this can be generalized.  Instead of location_key_match, could we have update.py take in a class name, list of class names, or regex of class names to run (such as pipelines.epidemiology.at_authority.AustriaLevel3DataSource) ?

Comment (owahltinez) [reply src/pipelines/epidemiology/config.yaml:120]: Filtering by class name is a great idea, and it would be a good addition. I don't think it solves the use case I had in mind, which is: what is the data for location `X` after these changes? Adding all the classes by hand sounds very error prone.

Say that we parsed each `*` as `[^_].+` and sandwiched the expression between `^` and `$`. Then `AT_*` would become `^AT_[^_]+$`, which should only match level 1. And `AT_*_*` would become `^AT_[^_]+_[^_]+$` which should only match level 2.

Even though this would depend on data sources providing the right regex, we can enforce that via unit tests. Currently we check if data sources output at least one row, we can easily add [another test](https://github.com/GoogleCloudPlatform/covid-19-open-data/blob/main/src/test/test_source_run.py#L89) to check that all output location keys match what's in the test string. Keep in mind that data sources can provide a list of regex, not just a single value ([example](https://github.com/GoogleCloudPlatform/covid-19-open-data/blob/main/src/pipelines/demographics/config.yaml#L109)).

If we get this to work well, we can re-add the old functionality which told us what outputs changed with each PR, since running a small subset of data sources would be much more feasible. For example, imagine adding a tag to the PR description like `loc:AT_*_*` which would trigger a data pipeline run using `--location-key AT_*_*` and attach the output to the PR discussion...

Comment (owahltinez) [reply src/pipelines/epidemiology/config.yaml:120]: I guess if we're going to replace `*` with `[^_]+` we could also do it for the test parameter in the `config.yaml`, since using plain regex is very powerful but also extremely error prone...

Comment (geening) [reply src/pipelines/epidemiology/config.yaml:120]: I think I was mistaken -- I've been referring to the country as level 1, state as 2 and district as 3, but I see here that state is 1 and district is 2 (so I suppose we're probably using 0 to denote country?): [https://screenshot.googleplex.com/5HwZobRsSdskyvg](https://screenshot.googleplex.com/5HwZobRsSdskyvg)

`^AT_[^_]+$` matches levels 1 and 2 but not the country level:


```
(base) mgl@mgl-macbookpro:~/src/covid-19-open-data/src$ python
Python 3.7.6 (default, Jan  8 2020, 13:42:34) 
[Clang 4.0.1 (tags/RELEASE_401/final)] :: Anaconda, Inc. on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import re
>>> re.match('^AT_[^_]+$', 'AT')
>>> re.match('^AT_[^_]+$', 'AT_1')
<re.Match object; span=(0, 4), match='AT_1'>
>>> re.match('^AT_[^_]+$', 'AT_101')
<re.Match object; span=(0, 6), match='AT_101'>

```

But good idea to sandwich the expression between `^` and `$`, and thanks for letting me know we can use a list of regex.  I don't think we need to parse `*` as `[^_].+`.  What if we just used the list [`AT`, `AT_.`] for the level 0/1 source and used `AT_...` for the level 2 source?

```
>>> re.match('^AT$', 'AT')
<re.Match object; span=(0, 2), match='AT'>
>>> re.match('^AT_.$', 'AT_1')
<re.Match object; span=(0, 4), match='AT_1'>
>>> re.match('^AT$', 'AT_101')
>>> re.match('^AT_.$', 'AT_101')
>>> re.match('^AT_...$', 'AT_101')
<re.Match object; span=(0, 6), match='AT_101'>

```

Comment (owahltinez) [reply src/pipelines/epidemiology/config.yaml:120]: > I think I was mistaken -- I've been referring to the country as level 1, state as 2 and district as 3, but I see here that state is 1 and district is 2 (so I suppose we're probably using 0 to denote country?)

Yes, we generally use the notation country level, admin level 1 (L1) and admin level 2 (L2). I think some of the comments and class names are not consistent with this...

> But good idea to sandwich the expression between ^ and $, and thanks for letting me know we can use a list of regex. I don't think we need to parse * as [^_].+. What if we just used the list [AT, AT_.] for the level 0/1 source and used AT_... for the level 2 source?

Your comment made me realize that the keys for AT admin level 2 are not correct. They need to be in the form `AT_${subregion1_code}_${subregion2_code}` (admin level 2 should *always* have two underscores). So, if the `subregion2_code` is `108`, then the key needs to be `AT_1_108`. If you think the `108` is unnecessary (since `1` should be obvious from `subregion1_code`) then you can change the `subregion2_code` to just the last 2 digits (i.e. key `AT_1_08`). I'll leave it up to you.

Comment (geening) [reply src/pipelines/epidemiology/config.yaml]: Changed the numbering of the administrative levels and fixed the level 2 keys.

Regarding the use case of "what is the data for location X after these changes", the current location_key_match still doesn't solve this use case for any regex X, and for example wouldn't be able to produce the data for Austria level 1 only but not the country-level, since they're both in the same source.  I suppose this isn't critical, though.

I'm hesitant to automatically replace `*` with `[^_]+` (and more weakly object to padding with `^` and `$`) because it creates a modified regex notation that can be confusing.  If keys are standard enough to support it, though, I do like the idea of  something like `update.py [...] --country XX --level [0,1,2,3]` and building the proper RegEx automatically without having to type it all out on the command line.  This still doesn't solve the problem of data sources providing data for multiple levels, though.

File: src/data/eurostat/index.csv
Comment (owahltinez) [inline src/data/eurostat/index.csv:22]: The keys in this file need updating too

**Review by owahltinez — CHANGES_REQUESTED** [review]

File: src/pipelines/epidemiology/config.yaml
Comment (owahltinez) [reply src/pipelines/epidemiology/config.yaml]: > Changed the numbering of the administrative levels and fixed the level 2 keys.

LGTM, you only forgot to update the keys in one of the auxiliary tables.

> I do like the idea of something like update.py [...] --country XX --level [0,1,2,3] and building the proper RegEx automatically without having to type it all out on the command line

Fair enough. If we go with that, we can do post-filtering of the data source like we are doing when the option `--strict-match` is passed.

File: src/data/eurostat/index.csv
Comment (geening) [reply src/data/eurostat/index.csv]: Updated, also updated the config.yaml so the Eurostat source runs for location-key AT_*_*

**Review by owahltinez — COMMENTED** [review]
LGTM, thanks Matt!

**Review by owahltinez — APPROVED** [review]

**Comment by geening** [comment]
Thanks Oscar!  I'll leave the English names of the provinces for another PR.

File: src/data/metadata.csv
Comment (geening) [reply src/data/metadata.csv]: There are only three districts where Wikidata's names are different for EN and DE, except for the "Bezirk" prefix versus "District" suffix:
```
                                     place                 label_de   label_en
77  http://www.wikidata.org/entity/Q694399                    Krems Krems-Land
97    http://www.wikidata.org/entity/Q1741                     Wien     Vienna
100  http://www.wikidata.org/entity/Q41753 Klagenfurt am Wörthersee Klagenfurt
```

Our metadata.csv already uses `Krems-Land` and `Klagenfurt Stadt` and I suppose it makes sense to keep these and not remove the `Stadt`, correct?  Should I change `Wien` to `Vienna`?

**Pull request merged**
